File: src/lib/parcelGeometry.ts
/**
 * Esri polygon rings to GeoJSON.
 *
 * ArcGIS exteriors are clockwise and holes are counter-clockwise. GeoJSON
 * (RFC 7946) wants the opposite. A positive shoelace area is a hole, not a
 * new exterior. Simplify stays near 1–7 meters so a shoreline is not replaced
 * by its first three vertices.
 */

/** ~1.3 m at Florida latitudes. */
export const PARCEL_SIMPLIFY_TOLERANCE = 0.000012;
/** ~4 m, used only when a ring is still very dense. */
export const PARCEL_SIMPLIFY_RELAXED = 0.000035;
/** ~7 m ceiling. The old seed jumped to ~30 m. */
export const PARCEL_SIMPLIFY_MAX = 0.00006;

export function signedArea(coords: number[][]): number {
  let area = 0;
  for (let i = 0; i < coords.length - 1; i += 1) {
    area += coords[i][0] * coords[i + 1][1] - coords[i + 1][0] * coords[i][1];
  }
  return area / 2;
}

function perpDist(p: [number, number], a: [number, number], b: [number, number]): number {
  const dx = b[0] - a[0];
  const dy = b[1] - a[1];
  if (dx === 0 && dy === 0) return Math.hypot(p[0] - a[0], p[1] - a[1]);
  const t = Math.max(0, Math.min(1, ((p[0] - a[0]) * dx + (p[1] - a[1]) * dy) / (dx * dx + dy * dy)));
  return Math.hypot(p[0] - (a[0] + t * dx), p[1] - (a[1] + t * dy));
}

function douglasPeucker(points: [number, number][], tol: number): [number, number][] {
  if (points.length <= 2) return points;
  const stack: Array<[number, number]> = [[0, points.length - 1]];
  const keep = new Set<number>([0, points.length - 1]);
  while (stack.length) {
    const [start, end] = stack.pop()!;
    let maxD = 0;
    let idx: number | null = null;
    for (let i = start + 1; i < end; i += 1) {
      const dist = perpDist(points[i], points[start], points[end]);
      if (dist > maxD) {
        maxD = dist;
        idx = i;
      }
    }
    if (idx != null && maxD > tol) {
      keep.add(idx);
      stack.push([start, idx], [idx, end]);
    }
  }
  return [...keep].sort((a, b) => a - b).map((i) => points[i]);
}

function close(points: [number, number][]): [number, number][] {
  if (!points.length) return points;
  const first = points[0];
  const last = points[points.length - 1];
  if (first[0] !== last[0] || first[1] !== last[1]) return [...points, first];
  return points;
}

function roundRing(points: [number, number][]): number[][] {
  return points.map(([x, y]) => [Math.round(x * 1e6) / 1e6, Math.round(y * 1e6) / 1e6]);
}

export function simplifyRing(coords: number[][], tol = PARCEL_SIMPLIFY_TOLERANCE): number[][] {
  if (coords.length < 4) return [];
  const open = coords[0][0] === coords[coords.length - 1][0] && coords[0][1] === coords[coords.length - 1][1]
    ? coords.slice(0, -1)
    : coords.slice();
  const pts = open.map(([x, y]) => [x, y] as [number, number]);
  if (pts.length < 3) return [];
  if (pts.length <= 4) return roundRing(close(pts));
  let simplified = douglasPeucker(pts, tol);
  if (simplified.length > 500) simplified = douglasPeucker(pts, Math.max(tol, PARCEL_SIMPLIFY_RELAXED));
  if (simplified.length > 800) simplified = douglasPeucker(pts, Math.max(tol, PARCEL_SIMPLIFY_MAX));
  if (simplified.length < 3) {
    const step = Math.max(1, Math.floor(pts.length / 80));
    simplified = pts.filter((_, index) => index % step === 0);
    if (simplified.length < 3) simplified = pts;
  }
  simplified = close(simplified);
  if (simplified.length < 4) simplified = close(pts);
  return roundRing(simplified);
}

function ensureClosed(raw: number[][]): number[][] {
  const first = raw[0];
  const last = raw[raw.length - 1];
  if (first[0] !== last[0] || first[1] !== last[1]) return [...raw, [...first]];
  return raw;
}

/** Group Esri rings into polygons and emit RFC 7946 winding. */
export function esriRingsToGeoJSON(
  rings: number[][][] | undefined,
  tol = PARCEL_SIMPLIFY_TOLERANCE,
): GeoJSON.Polygon | GeoJSON.MultiPolygon | null {
  const polygons: number[][][][] = [];
  let current: number[][][] = [];
  for (const ring of rings ?? []) {
    let raw = ring.map(([x, y]) => [x, y]);
    if (raw.length < 4) continue;
    raw = ensureClosed(raw);
    const area = signedArea(raw);
    if (Math.abs(area) < 1e-14) continue;
    const coords = simplifyRing(raw, tol);
    if (coords.length < 4) continue;
    const isHole = area > 0;
    if (isHole && current.length) current.push(coords);
    else {
      if (current.length) polygons.push(current);
      current = [coords];
    }
  }
  if (current.length) polygons.push(current);
  if (!polygons.length) return null;

  const normalized = polygons.map((poly) => {
    let outer = poly[0];
    if (signedArea(outer) < 0) outer = [...outer].reverse();
    const holes = poly.slice(1).map((hole) => (signedArea(hole) > 0 ? [...hole].reverse() : hole));
    return [outer, ...holes];
  });
  if (normalized.length === 1) return { type: "Polygon", coordinates: normalized[0] };
  return { type: "MultiPolygon", coordinates: normalized };
}

function pointInRing(x: number, y: number, ring: number[][]): boolean {
  let inside = false;
  let j = ring.length - 1;
  for (let i = 0; i < ring.length; i += 1) {
    const xi = ring[i][0];
    const yi = ring[i][1];
    const xj = ring[j][0];
    const yj = ring[j][1];
    if (yi > y !== yj > y && x < ((xj - xi) * (y - yi)) / (yj - yi || 1e-12) + xi) inside = !inside;
    j = i;
  }
  return inside;
}

function insidePolygon(x: number, y: number, poly: number[][][]): boolean {
  if (!poly.length || !pointInRing(x, y, poly[0])) return false;
  return poly.slice(1).every((hole) => !pointInRing(x, y, hole));
}

function partsOf(geometry: GeoJSON.Polygon | GeoJSON.MultiPolygon): number[][][][] {
  return geometry.type === "Polygon" ? [geometry.coordinates] : geometry.coordinates;
}

/** A point inside the largest part. Vertex averages fall outside concave rings. */
export function representativePoint(geometry: GeoJSON.Polygon | GeoJSON.MultiPolygon): [number, number] | null {
  let best: number[][][] | null = null;
  let bestArea = -1;
  for (const poly of partsOf(geometry)) {
    if (!poly[0]) continue;
    const area = Math.abs(signedArea(poly[0]));
    if (area > bestArea) {
      bestArea = area;
      best = poly;
    }
  }
  if (!best) return null;
  const ring = best[0];
  const body = ring.slice(0, -1);
  if (!body.length) return null;
  const lon = body.reduce((sum, p) => sum + p[0], 0) / body.length;
  const lat = body.reduce((sum, p) => sum + p[1], 0) / body.length;
  const round = (x: number, y: number): [number, number] => [Math.round(x * 1e6) / 1e6, Math.round(y * 1e6) / 1e6];
  if (insidePolygon(lon, lat, best)) return round(lon, lat);
  const xs = body.map((p) => p[0]);
  const ys = body.map((p) => p[1]);
  const bx = (Math.min(...xs) + Math.max(...xs)) / 2;
  const by = (Math.min(...ys) + Math.max(...ys)) / 2;
  if (insidePolygon(bx, by, best)) return round(bx, by);
  const step = Math.max(1, Math.floor(body.length / 16));
  for (let i = 0; i < body.length; i += step) {
    const qlon = body[i][0] * 0.8 + lon * 0.2;
    const qlat = body[i][1] * 0.8 + lat * 0.2;
    if (insidePolygon(qlon, qlat, best)) return round(qlon, qlat);
  }
  return round(body[0][0], body[0][1]);
}

File: scripts/parcel_geometry.py
"""Esri polygon rings to GeoJSON, with a cadastral simplify tolerance.

ArcGIS exterior rings are clockwise. Holes are counter-clockwise.
GeoJSON (RFC 7946) wants the opposite winding. Treating a positive shoelace
area as a new exterior turns holes into extra polygons and later exteriors
into holes. That is the garbage-outline bug.

Simplify stays near 1–7 meters. The old path jumped to ~30 meters on dense
shorelines and, when that collapsed, kept the first three vertices.
"""

from __future__ import annotations

import math

# ~1.3 m, ~4 m, ~7 m at Florida latitudes. Do not jump to ~30 m.
BASE_TOL = 0.000012
RELAX_TOL = 0.000035
MAX_TOL = 0.00006


def signed_area(coords: list[list[float]]) -> float:
    area = 0.0
    for i in range(len(coords) - 1):
        area += coords[i][0] * coords[i + 1][1] - coords[i + 1][0] * coords[i][1]
    return area / 2.0


def ring_signed_m2(coords: list[list[float]]) -> float:
    """Signed area in square meters. Clockwise ArcGIS rings are negative."""
    if len(coords) < 4:
        return 0.0
    lat0 = sum(p[1] for p in coords) / len(coords)
    kx = 111_320 * math.cos(math.radians(lat0))
    ky = 110_540
    area = 0.0
    for i in range(len(coords) - 1):
        x1, y1 = coords[i][0] * kx, coords[i][1] * ky
        x2, y2 = coords[i + 1][0] * kx, coords[i + 1][1] * ky
        area += x1 * y2 - x2 * y1
    return area / 2.0


def perp_dist(p: tuple[float, float], a: tuple[float, float], b: tuple[float, float]) -> float:
    ax, ay = a
    bx, by = b
    px, py = p
    dx, dy = bx - ax, by - ay
    if dx == 0 and dy == 0:
        return math.hypot(px - ax, py - ay)
    t = ((px - ax) * dx + (py - ay) * dy) / (dx * dx + dy * dy)
    t = max(0.0, min(1.0, t))
    return math.hypot(px - (ax + t * dx), py - (ay + t * dy))


def douglas_peucker(points: list[tuple[float, float]], tol: float) -> list[tuple[float, float]]:
    if len(points) <= 2:
        return points
    stack = [(0, len(points) - 1)]
    keep = {0, len(points) - 1}
    while stack:
        start, end = stack.pop()
        max_d = 0.0
        idx = None
        for i in range(start + 1, end):
            dist = perp_dist(points[i], points[start], points[end])
            if dist > max_d:
                max_d = dist
                idx = i
        if idx is not None and max_d > tol:
            keep.add(idx)
            stack.append((start, idx))
            stack.append((idx, end))
    return [points[i] for i in sorted(keep)]


def _close(points: list[tuple[float, float]]) -> list[tuple[float, float]]:
    if not points:
        return points
    if points[0] != points[-1]:
        return points + [points[0]]
    return points


def _round_ring(points: list[tuple[float, float]]) -> list[list[float]]:
    return [[round(x, 6), round(y, 6)] for x, y in points]


def simplify_ring(coords: list[list[float]], tol: float = BASE_TOL) -> list[list[float]]:
    """Simplify one ring. Never replace a real ring with its first three vertices."""
    if len(coords) < 4:
        return []
    open_ring = coords[:-1] if coords[0] == coords[-1] else list(coords)
    pts = [(float(x), float(y)) for x, y in open_ring]
    if len(pts) < 3:
        return []
    if len(pts) <= 4:
        return _round_ring(_close(pts))
    simplified = douglas_peucker(pts, tol)
    if len(simplified) > 500:
        simplified = douglas_peucker(pts, max(tol, RELAX_TOL))
    if len(simplified) > 800:
        simplified = douglas_peucker(pts, max(tol, MAX_TOL))
    if len(simplified) < 3:
        step = max(1, len(pts) // 80)
        simplified = pts[::step]
        if len(simplified) < 3:
            simplified = pts
    simplified = _close(simplified)
    if len(simplified) < 4:
        simplified = _close(pts)
    return _round_ring(simplified)


def _ensure_closed(raw: list[list[float]]) -> list[list[float]]:
    if raw[0] != raw[-1]:
        return raw + [raw[0]]
    return raw


def esri_rings_to_geojson(rings: list, tol: float = BASE_TOL) -> dict | None:
    """Group Esri rings into polygons and emit RFC 7946 winding."""
    polygons: list[list[list[list[float]]]] = []
    current: list[list[list[float]]] = []
    for ring in rings or []:
        raw = [[float(x), float(y)] for x, y in ring]
        if len(raw) < 4:
            continue
        raw = _ensure_closed(raw)
        area = signed_area(raw)
        if abs(area) < 1e-14:
            continue
        coords = simplify_ring(raw, tol)
        if len(coords) < 4:
            continue
        # Positive shoelace is counter-clockwise: an Esri hole.
        is_hole = area > 0
        if is_hole and current:
            current.append(coords)
        else:
            if current:
                polygons.append(current)
            current = [coords]
    if current:
        polygons.append(current)
    if not polygons:
        return None

    normalized: list[list[list[list[float]]]] = []
    for poly in polygons:
        outer = poly[0]
        if signed_area(outer) < 0:
            outer = list(reversed(outer))
        holes: list[list[list[float]]] = []
        for hole in poly[1:]:
            if signed_area(hole) > 0:
                hole = list(reversed(hole))
            holes.append(hole)
        normalized.append([outer, *holes])
    if len(normalized) == 1:
        return {"type": "Polygon", "coordinates": normalized[0]}
    return {"type": "MultiPolygon", "coordinates": normalized}


def polygon_parts(geometry: dict) -> list:
    if geometry.get("type") == "Polygon":
        return [geometry["coordinates"]]
    if geometry.get("type") == "MultiPolygon":
        return list(geometry["coordinates"])
    return []


def _point_in_ring(x: float, y: float, ring: list[list[float]]) -> bool:
    inside = False
    j = len(ring) - 1
    for i in range(len(ring)):
        xi, yi = ring[i]
        xj, yj = ring[j]
        if ((yi > y) != (yj > y)) and (x < (xj - xi) * (y - yi) / ((yj - yi) or 1e-12) + xi):
            inside = not inside
        j = i
    return inside


def _inside_polygon(x: float, y: float, poly: list[list[list[float]]]) -> bool:
    if not poly or not _point_in_ring(x, y, poly[0]):
        return False
    return not any(_point_in_ring(x, y, hole) for hole in poly[1:])


def contains_point(geometry: dict | None, x: float, y: float) -> bool:
    """True when (x, y) is inside an exterior ring and outside holes."""
    if not geometry:
        return False
    return any(_inside_polygon(x, y, poly) for poly in polygon_parts(geometry))


def point_in_geometry(lon: float, lat: float, geometry: dict | None) -> bool:
    if not geometry:
        return False
    return any(_inside_polygon(lon, lat, poly) for poly in polygon_parts(geometry))


def representative_point(geometry: dict) -> tuple[float, float] | None:
    """A point inside the largest part. Vertex averages fall outside concave rings."""
    parts = polygon_parts(geometry)
    best = None
    best_area = -1.0
    for poly in parts:
        if not poly or not poly[0]:
            continue
        area = abs(signed_area(poly[0]))
        if area > best_area:
            best_area = area
            best = poly
    if not best:
        return None
    ring = best[0]
    body = ring[:-1] if len(ring) > 1 else ring
    if not body:
        return None
    lon = sum(p[0] for p in body) / len(body)
    lat = sum(p[1] for p in body) / len(body)
    if _inside_polygon(lon, lat, best):
        return round(lon, 6), round(lat, 6)
    xs = [p[0] for p in body]
    ys = [p[1] for p in body]
    bx = (min(xs) + max(xs)) / 2
    by = (min(ys) + max(ys)) / 2
    if _inside_polygon(bx, by, best):
        return round(bx, 6), round(by, 6)
    step = max(1, len(body) // 16)
    for point in body[::step]:
        qlon = point[0] * 0.8 + lon * 0.2
        qlat = point[1] * 0.8 + lat * 0.2
        if _inside_polygon(qlon, qlat, best):
            return round(qlon, 6), round(qlat, 6)
    return round(body[0][0], 6), round(body[0][1], 6)


def net_acres(rings: list) -> float:
    net = 0.0
    for ring in rings or []:
        raw = [[float(x), float(y)] for x, y in ring]
        if len(raw) < 4:
            continue
        if raw[0] != raw[-1]:
            raw = raw + [raw[0]]
        net += ring_signed_m2(raw)
    return abs(net) / 4046.8564224


def planar_acres(rings: list) -> float:
    """Net acres from planar feet. Georgia West State Plane uses US survey feet."""
    net = 0.0
    for ring in rings or []:
        raw = [[float(x), float(y)] for x, y in ring]
        if len(raw) < 4:
            continue
        if raw[0] != raw[-1]:
            raw = raw + [raw[0]]
        net += signed_area(raw)
    return abs(net) / 43560.0


# NAD83 / Georgia West (ftUS). ESRI 102667, EPSG:2240.
_A = 6378137.0
_F = 1 / 298.257222101
_E2 = 2 * _F - _F * _F
_EP2 = _E2 / (1 - _E2)
_K0 = 0.9999
_LON0 = math.radians(-84.16666666666667)
_LAT0 = math.radians(30.0)
_FALSE_E = 2296583.333333333
_FALSE_N = 0.0
_FT_US = 1200 / 3937
_E1 = (1 - math.sqrt(1 - _E2)) / (1 + math.sqrt(1 - _E2))
_M_DENOM = _A * (1 - _E2 / 4 - 3 * _E2**2 / 64 - 5 * _E2**3 / 256)


def _meridional_arc(phi: float) -> float:
    e2 = _E2
    e4 = e2 * e2
    e6 = e4 * e2
    a0 = 1 - e2 / 4 - 3 * e4 / 64 - 5 * e6 / 256
    a2 = 3 / 8 * (e2 + e4 / 4 + 15 * e6 / 128)
    a4 = 15 / 256 * (e4 + 3 * e6 / 4)
    a6 = 35 * e6 / 3072
    return _A * (a0 * phi - a2 * math.sin(2 * phi) + a4 * math.sin(4 * phi) - a6 * math.sin(6 * phi))


_M0 = _meridional_arc(_LAT0)


def ga_west_ft_to_wgs84(east: float, north: float) -> tuple[float, float]:
    """Inverse transverse Mercator from Georgia West US survey feet to WGS84 degrees."""
    x = (east - _FALSE_E) * _FT_US
    y = (north - _FALSE_N) * _FT_US
    mu = (_M0 + y / _K0) / _M_DENOM
    e1 = _E1
    phi1 = (
        mu
        + (3 * e1 / 2 - 27 * e1**3 / 32) * math.sin(2 * mu)
        + (21 * e1**2 / 16 - 55 * e1**4 / 32) * math.sin(4 * mu)
        + (151 * e1**3 / 96) * math.sin(6 * mu)
        + (1097 * e1**4 / 512) * math.sin(8 * mu)
    )
    cos1 = math.cos(phi1)
    sin1 = math.sin(phi1)
    c1 = _EP2 * cos1 * cos1
    t1 = math.tan(phi1) ** 2
    n1 = _A / math.sqrt(1 - _E2 * sin1 * sin1)
    r1 = _A * (1 - _E2) / (1 - _E2 * sin1 * sin1) ** 1.5
    d = x / (n1 * _K0)
    d2 = d * d
    lat = phi1 - (n1 * math.tan(phi1) / r1) * (
        d2 / 2
        - (5 + 3 * t1 + 10 * c1 - 4 * c1 * c1 - 9 * _EP2) * d2 * d2 / 24
        + (61 + 90 * t1 + 298 * c1 + 45 * t1 * t1 - 252 * _EP2 - 3 * c1 * c1) * d2 * d2 * d2 / 720
    )
    lon = _LON0 + (
        d
        - (1 + 2 * t1 + c1) * d2 * d / 6
        + (5 - 2 * c1 + 28 * t1 - 3 * c1 * c1 + 8 * _EP2 + 24 * t1 * t1) * d2 * d2 * d / 120
    ) / cos1
    return math.degrees(lon), math.degrees(lat)


def project_ga_west_rings(rings: list) -> list:
    projected = []
    for ring in rings or []:
        projected.append([list(ga_west_ft_to_wgs84(float(x), float(y))) for x, y in ring])
    return projected


def geometry_contains(geometry: dict, x: float, y: float) -> bool:
    return any(_inside_polygon(x, y, poly) for poly in polygon_parts(geometry))
